fvm/src/kernel/default.rs

```diff
@@ -436,34 +435,28 @@ impl<C> CryptoOps for DefaultKernel<C>
 where
     C: CallManager,
 {
-    fn verify_signature(
+    fn verify_bls_aggregate(
         &self,
-        sig_type: SignatureType,
-        signature: &[u8],
-        signer: &Address,
-        plaintext: &[u8],
+        aggregate_sig: &[u8; BLS_SIG_LEN],
+        pub_keys: &[&[u8; BLS_PUB_LEN]],
+        digests: &[&[u8; BLS_DIGEST_LEN]],
     ) -> Result<bool> {
         let t = self.call_manager.charge_gas(
             self.call_manager
                 .price_list()
-                .on_verify_signature(sig_type, plaintext.len()),
+                .on_verify_aggregate_signature(pub_keys.len()),
         )?;
 
-        // We only support key addresses (f1/f3). This change does not require a FIP, because no
-        // actors invoke this method with non-key addresses.
-        let signing_addr = match signer.payload() {
-            Payload::BLS(_) | Payload::Secp256k1(_) => *signer,
-            // Not a key address.
-            _ => {
-                return Err(syscall_error!(IllegalArgument; "address protocol {} not supported", signer.protocol()).into());
-            }
-        };
-
-        // Verify signature, catching errors. Signature verification can include some complicated
-        // math.
-        t.record(catch_and_log_panic("verifying signature", || {
-            Ok(signature::verify(sig_type, signature, plaintext, &signing_addr).is_ok())
-        }))
+        t.record(
+            signature::ops::verify_bls_aggregate(aggregate_sig, pub_keys, digests).map_err(|e| {
```

Comment:
We'd generally just map all errors here to "false", reserving "illegal argument" errors for cases where, e.g., the caller passed arguments that just don't make sense. Is there any way to distinguish between the two?

E.g., maybe first (before we even charge for gas) check if `pub_keys.len() == digests.len()` and return an illegal argument in that case. Otherwise, verify the signature and return `false` for any error case.


For context, "illegal argument" usually means that there's a bug in the smart contract somewhere (i.e., it's the smart contract's fault).

Reply:
`shared`'s `verify_bls_aggregate` returns an `Err(String)` if any of the byte array arguments (sig, pub keys, digests) represent an invalid curve point or if `pub_keys.len() != digests.len()`. I returned `IllegalArgument` because those were user-supplied errors, but I can also just map those to `false`. The `shared` function returns `Ok(false)` on signature verification failure (including if some of `bls_signatures` security requirements are unmet, e.g. repeated digests being aggregated, or single signer aggregation with the public key `0`).

Reply:
I would say: invalid curve points are just "invalid signature". Technically, the public keys having invalid curve points could be considered an "invalid argument", but that makes documenting the invalid argument error case more difficult so I'm going to optimize for simpler documentation...

Reply:
Sounds good. Also, parsing the `String` error returned by `shared::[...]::verify_bls_aggregate` for exclusively curve point errors could get kludgy.

Reply:
Yeah, let's keep it simple. Anything else risks causing forks.